deepchem/hyper/base_classes.py

```diff
@@ -0,0 +1,83 @@
+class HyperparamOpt(object):
+  """Abstract superclass for hyperparameter search classes.
+
+  This class is an abstract base class for hyperparameter search
+  classes in DeepChem. Hyperparameter search is performed on
+  `dc.model.Model` classes. Each hyperparameter object accepts a
+  `dc.models.Model` class upon construct. When the `hyperparam_search`
+  class is invoked, this class is used to construct many different
+  concrete models which are trained on the specified training set and
+  evaluated on a given validation set.
+
+  Different subclasses of `HyperparamOpt` differ in the choice of
+  strategy for searching the hyperparameter evaluation space. This
+  class itself is an abstract superclass and should never be directly
+  instantiated.
+  """
+
+  def __init__(self, model_builder):
+    """Initialize Hyperparameter Optimizer.
+
+    Note this is an abstract constructor which should only be used by
+    subclasses.
+
+    Parameters
+    ----------
+    model_builder: constructor function.
+      This parameter must be constructor function which returns an
+      object which is an instance of `dc.models.Model`. This function
+      must accept two arguments, `model_params` of type `dict` and
+      `model_dir`, a string specifying a path to a model directory.
+      See the example.
+    """
+    if self.__class__.__name__ == "HyperparamOpt":
+      raise ValueError(
+          "HyperparamOpt is an abstract superclass and cannot be directly instantiated. You probably want to instantiate a concrete subclass instead."
+      )
+    self.model_builder = model_builder
+
+  def hyperparam_search(self,
+                        params_dict,
+                        train_dataset,
+                        valid_dataset,
+                        transformers,
+                        metric,
+                        use_max=True,
+                        logdir=None):
+    """Conduct Hyperparameter search.
+
+    This method defines the common API shared by all hyperparameter
+    optimization subclasses. Different classes will implement
+    different search methods but they must all follow this common API.
+
+    Parameters
+    ----------
+    params_dict: dict
+      Dictionary mapping strings to Ints/Floats/Lists. Note that the
```

Comment:
Are those really the only allowed types?  What if a model takes arguments of a different type, like a string or a bool?  It needs to explain the significance of those types and how this parameter is interpreted.

Reply:
Good point! I've fixed this and added some more clarifying documentation